rfcs/0042-config-option.md

```diff
@@ -0,0 +1,191 @@
+
+### Documentation
+
+The nixpkgs manual should be updated to recommend this way of doing program configuration in modules, along with examples.
+
+## Limitations
+
+- Limited to configuration file formats representable conveniently in Nix, such as JSON, YAML, INI, key-value files, or similar formats. Examples of unsuitable configuration formats are Haskell, Lisp, Lua or other generic programming languages. For those it is recommended to not hardcode anything and provide a `config` option with type `types.str` or `types.lines` if it makes sense to merge multiple assignments of it.
+
+## Additional config options
+Sometimes it makes sense to have an additional NixOS option for a specific configuration setting. In general this should be discussed on a case-by-case basis to judge whether it makes sense. However keep in mind that it's always possible to add more options later on, but you can't as easily remove existing options. Instances of where it can make sense are:
+- Settings that are necessary for the module to work and are different for every user, so they can't have a default. Examples are `services.dd-agent.api_key`, `services.davmail.url` or `services.hydra.hydraURL`. Having a separate option for these settings can give a much better error message when you don't set them (instead of failing at runtime or having to encode the requirement in an assertion) and better discoverability.
+- Password settings: Some programs configuration files require passwords in them directly. Since we try to avoid having passwords in the Nix store, it is advisable to provide a `passwordFile` option as well, which would replace a placeholder password in the configuration file at runtime.
+
+This `config` approach described here is very flexible for these kind of additions, here's an example where we add a `domain` setting to the above service as a separate NixOS option:
+```nix
+{ config, lib, ... }: with lib; {
+
+  options.services.foo = {
+    # ...
+    
+    domain = mkOption {
+      type = types.str;
+      description = "Domain this service operates on.";
+    };
+  };
+  
+  config = mkIf cfg.enable {
+    services.foo.config = {
+      # ...
+      domain = mkDefault cfg.domain;
+    };
+  };
+}
+```
+
+Even though we have two ways of specifying this configuration setting now, the user is free to choose either way.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are some disadvantages to this approach:
+- Types of configuration settings can't be checked*, which can lead to packages failing at runtime instead of evaluation time. If the program supports a mode for checking its configuration, this problem can be solved elegantly by using it.
```

Comment:
As such problems can be fairly painful (i.e. when screwing up SSH configs), I'd suggest that we actually add a recommendation to the actual RFC to use config checking tools (if possible) rather than having  a simple note at the end :)

Reply:
Good idea, will do